### PR TITLE
Fixes bugs reported by Clang Static Analyzer.

### LINKED ISSUE #20: bugs found by llvm static analyzer
File:	ptp.c
Warning:	line 2012, column 10
Result of 'calloc' is converted to a pointer of type 'unsigned char', which is incompatible with sizeof operand type;  'uint32_t'data = calloc(sizeof(uint32_t),3);

File:	./ptp-pack.c
Warning:	line 252, column 10
Result of 'malloc' is converted to a pointer of type 'unsigned char', which is incompatible with sizeof operand type 'uint32_t' : *data = malloc ((arraylen+1)*sizeof(uint32_t));

### PR DESCRIPTION
**Pull request by claire50**
Fixes #20.

### CODE REVIEW

File: plugins/ptp/libptp/ptp.c [hunk 1]
```diff
@@ -277,6 +277,8 @@ ptp_init_send_memory_handler(PTPDataHandler *handler,
 // free private struct + data
 static uint16_t
 ptp_exit_send_memory_handler (PTPDataHandler *handler) {
+	if (!handler->priv)
+		return PTP_RC_OK;
```
Comment (owenca) [inline plugins/ptp/libptp/ptp.c]: Undo the above change.

This is a tricky one, as the execution path consists of 15 jumps. Anyway, trying to fix the bug at the end of the path won't work here. The best place to tackle the problem is much earlier, before Step 4 on the path:
Move Line 268 to right before Line 266.

File: plugins/ptp/libptp/usb.c [hunk 2]
```diff
@@ -164,6 +164,7 @@ ptp_usb_senddata (PTPParams* params, PTPContainer* ptp,
 	res = ptp_write_func((char*)&usbdata, wlen, params->data);
 	if (res != wlen) 
 	{
+		ret = PTP_ERROR_IO;
```
Comment (owenca) [inline plugins/ptp/libptp/usb.c]:
```suggestion
```

File: plugins/ptp/libptp/usb.c [hunk 3]
```diff
@@ -164,6 +164,7 @@ ptp_usb_senddata (PTPParams* params, PTPContainer* ptp,
...
 		ptp_usb_error(params, "ptp2/usb_senddata : request code 0x%04x sending data error 0x%04x",ptp->Code,ret);
```
Comment (owenca) [inline plugins/ptp/libptp/usb.c]:
```suggestion
		ptp_usb_error(params, "ptp2/usb_senddata : request code 0x%04x sending data error 0x%04x",ptp->Code,PTP_ERROR_IO);
```
This is a little bit simpler. :)

**Comment by owenca** [comment]
https://github.com/HaikuArchives/PhotoGrabber/blob/f900c57b9b31b769514421b3b6cfdc0b77cdbb8a/plugins/ptp/libptp/ptp.c#L2014-L2015
Although this is a false positive, we can improve the code here and at the same time silence Clang Static Analyzer:
```
size = sizeof(uint32_t) * 3;
data = calloc(size, 1);
```

**Comment by owenca** [comment]
^I was talking about usb.c line 133.

File: plugins/ptp/libptp/ptp.c [hunk 4]
```diff
@@ -262,10 +262,10 @@ ptp_init_send_memory_handler(PTPDataHandler *handler,
 	unsigned char *data, unsigned long len
 ) {
 	PTPMemHandlerPrivate* priv;
+	handler->priv = priv;
```
Comment (korli) [inline plugins/ptp/libptp/ptp.c]: This doesn't make sense. priv is undefined at this point.

File: plugins/ptp/libptp/usb.c [hunk 5]
```diff
@@ -164,7 +164,7 @@ ptp_usb_senddata (PTPParams* params, PTPContainer* ptp,
 	res = ptp_write_func((char*)&usbdata, wlen, params->data);
 	if (res != wlen) 
 	{
-		ptp_usb_error(params, "ptp2/usb_senddata : request code 0x%04x sending data error 0x%04x",ptp->Code,ret);
+		ptp_usb_error(params, "ptp2/usb_senddata : request code 0x%04x sending data error 0x%04x", ptp->Code, PTP_ERROR_IO);
```
Comment (korli) [inline plugins/ptp/libptp/usb.c]: "sending data error" is actually "res".

File: plugins/ptp/libptp/ptp.c (hunk 4 above)
Comment (owenca) [reply plugins/ptp/libptp/ptp.c]: Yeah, the two lines should be swapped.

Comment (claire50) [reply plugins/ptp/libptp/ptp.c]: @korli @owenca I must have misunderstood Owen's comment. Now I see the whole point of the recommended fix.

File: plugins/ptp/libptp/usb.c (hunk 5 above)
Comment (claire50) [reply plugins/ptp/libptp/usb.c]: I get it! As a matter of fact, there is already a use case in the code:
https://github.com/HaikuArchives/PhotoGrabber/blob/master/plugins/ptp/libptp/usb.c#L123

**Pull request merged**